File: common/src/main/kotlin/net/impleri/playerskills/skills/specialized/SpecializedSkillType.kt
package net.impleri.playerskills.skills.specialized

import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.api.SkillType
import net.impleri.playerskills.utils.PlayerSkillsLogger
import net.impleri.playerskills.utils.SkillResourceLocation
import net.minecraft.resources.ResourceLocation

class SpecializedSkillType : SkillType<String>() {
  override val name: ResourceLocation
    get() = NAME

  private fun getOptions(skill: Skill<String>): List<String> {
    return skill.options
  }

  override fun castToString(value: String?): String? {
    return value
  }

  public override fun castFromString(value: String?): String? {
    return value
  }

  override fun can(skill: Skill<String>, expectedValue: String?): Boolean {
    if (expectedValue == null) {
      return skill.value != null
    }
    PlayerSkillsLogger.SKILLS.debug("Checking if player can ${skill.name} (is ${skill.value} == $expectedValue)")
    return expectedValue == skill.value
  }

  override fun getPrevValue(skill: Skill<String>, min: String?, max: String?): String {
    return ""
  }

  override fun getNextValue(skill: Skill<String>, min: String?, max: String?): String {
    return ""
  }

  companion object {
    var NAME = SkillResourceLocation.of("specialized")
  }
}

File: common/src/main/scala/net/impleri/playerskills/skills/specialized/SpecializedSkillType.scala
package net.impleri.playerskills.skills.specialized

import net.impleri.playerskills.api.skills.Skill
import net.impleri.playerskills.api.skills.SkillOps
import net.impleri.playerskills.api.skills.SkillType
import net.impleri.playerskills.utils.SkillResourceLocation
import net.minecraft.resources.ResourceLocation

case class SpecializedSkillType(override val skillOps: SkillOps = Skill()) extends SkillType[String] {
  override val name: ResourceLocation = SpecializedSkillType.NAME

  override def castToString(value: String): Option[String] = Option(value)

  override def castFromString(value: String): Option[String] = Option(value)

  override def can(skill: Skill[String], threshold: Option[String]): Boolean = {
    (skill.value, threshold) match {
      case (Some(v), Some(t)) => v == t
      case (Some(_), None) => true
      case _ => false
    }
  }

  override def getPrevValue(skill: Skill[String], min: Option[String], max: Option[String]): Option[String] = None

  override def getNextValue(skill: Skill[String], min: Option[String], max: Option[String]): Option[String] = None
}

object SpecializedSkillType {
  val NAME: ResourceLocation = SkillResourceLocation.of("specialized").get
}
